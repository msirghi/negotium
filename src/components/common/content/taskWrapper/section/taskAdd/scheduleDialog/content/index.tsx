export { MenuContent } from './container/MenuContent';
