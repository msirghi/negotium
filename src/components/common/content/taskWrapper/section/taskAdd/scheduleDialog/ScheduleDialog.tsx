import { FC, useState, MouseEvent } from 'react';
import { Menu } from '@mui/material';
import { MenuContent } from './content';
import DateUtils from '../../../../../../../common/utils/dateUtils';
import colors from '../../../../../../../common/styles/colors';
import { makeStyles } from '@mui/styles';
import { NullableDate } from '../../../../../../../common/types/common.types';

type Props = {
  onDateSelect: (date: NullableDate) => void;
  defaultDate?: Date;
};

const useStyles = makeStyles({
  root: {
    marginTop: '20px',
    borderRadius: 5,
    padding: '3px 5px',
    width: 'fit-content',
    cursor: 'pointer',
  },
});

export const ScheduleDialog: FC<Props> = ({ onDateSelect, defaultDate }) => {
  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
  const open = Boolean(anchorEl);
  const [selectedDate, setSelectedDate] = useState<NullableDate>(
    (defaultDate || null) as unknown as NullableDate
  );
  const classes = useStyles();

  const onDatePick = (date: NullableDate) => {
    onDateSelect(date);
    setSelectedDate(date);
    handleClose();
  };

  const handleClickListItem = (event: MouseEvent<HTMLElement>) => {
    setAnchorEl(event.currentTarget);
  };

  const handleClose = () => {
    setAnchorEl(null);
  };

  return (
    <div>
      <div
        id={'schedule-button'}
        onClick={handleClickListItem}
        className={classes.root}
        style={{
          border: selectedDate
            ? `1px solid ${colors.primaries.lightBlue_1}`
            : '1px solid grey',

          color: selectedDate ? colors.primaries.lightBlue_1 : '',
        }}
      >
        <div>
          {selectedDate ? DateUtils.getDateLabel(selectedDate) : 'Schedule'}
        </div>
      </div>
      <Menu anchorEl={anchorEl} open={open} onClose={handleClose}>
        <MenuContent onDatePick={onDatePick} selectedDate={selectedDate} />
      </Menu>
    </div>
  );
};
