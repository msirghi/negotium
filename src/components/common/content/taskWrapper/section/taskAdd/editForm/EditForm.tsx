import { FC, useState } from 'react';
import { ScheduleDialog } from '../scheduleDialog/ScheduleDialog';
import { Nullable } from '../../../../../../../common/types/common.types';
import { makeStyles } from '@mui/styles';
import MentionInput from '../../../../../form/input/mention/MentionInput';
import { MENTION_ARRAY_KEYWORDS } from '../../../../../../../common/constants/constants';
import SlateUtils from '../../../../../../../common/utils/slateUtils';
import FeatureToggles from '../../../../../../../utilities/featureToggles/FeatureToggles';
import { If } from '../../../../../utilities/if/If';
import { TextField } from '@mui/material';
import StringUtils from '../../../../../../../common/utils/stringUtils';

type Props = {
  fieldValue: string;
  setFieldValue: (title: string) => void;
  onDateSelect: (date: Nullable<Date>) => void;
  defaultDate?: Date;
};

const useStyles = makeStyles({
  root: {
    padding: 10,
    border: '1px solid lightgrey',
    borderRadius: 5,
  },
});

export const EditForm: FC<Props> = ({
  fieldValue,
  setFieldValue,
  onDateSelect,
  defaultDate,
}) => {
  const classes = useStyles();
  const [date, setDate] = useState<string>();
  const isSlateInputEnabled = FeatureToggles.isFeatureEnabled(
    FeatureToggles.keys.SLATE_INPUT
  );

  const onChange = (val: any) => {
    if (isSlateInputEnabled) {
      const stringified = JSON.stringify(val);
      setFieldValue(stringified);
      SlateUtils.detectDateKeywords(stringified, (date) => {
        setDate(date);
        onDateSelect(date as unknown as Nullable<Date>);
      });
      return;
    }

    const values = StringUtils.getTaskInputDateByKeywords(val.target.value);
    if (values.date) {
      setDate(values.date);
      onDateSelect(values.date as unknown as Nullable<Date>);
    }

    setFieldValue(values.value);
  };

  return (
    <div className={classes.root}>
      <If condition={isSlateInputEnabled}>
        <MentionInput onChange={onChange} keywords={MENTION_ARRAY_KEYWORDS} />
      </If>
      <If condition={!isSlateInputEnabled}>
        <TextField
          value={fieldValue}
          fullWidth
          size={'small'}
          variant={'standard'}
          inputProps={{
            'data-testid': 'tab-title-field',
          }}
          placeholder={'New task'}
          InputProps={{ disableUnderline: true }}
          onChange={onChange}
        />
      </If>
      <ScheduleDialog
        onDateSelect={onDateSelect}
        defaultDate={defaultDate}
        value={date as string}
      />
    </div>
  );
};
