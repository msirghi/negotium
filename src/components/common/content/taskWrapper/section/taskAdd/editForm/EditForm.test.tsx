import renderer from 'react-test-renderer';
import { EditForm } from './EditForm';
import { act, fireEvent, render } from '@testing-library/react';
import { mount } from 'enzyme';
import { ScheduleDialog } from '../scheduleDialog/ScheduleDialog';

describe('EditForm', () => {
  const defaultProps = {
    fieldValue: 'val',
    setFieldValue: jest.fn(),
    onDateSelect: jest.fn(),
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match the snapshot', () => {
    const tree = renderer.create(<EditForm {...defaultProps} />);
    expect(tree).toMatchSnapshot();
  });

  it('should call prop method on input change', () => {
    const { getByTestId } = render(<EditForm {...defaultProps} />);
    const input = getByTestId('tab-title-field');

    act(() => {
      fireEvent.change(input, { target: { value: 'new' } });
    });
    expect(defaultProps.setFieldValue).toBeCalled();
  });

  it('should call prop method on date select', () => {
    const wrapper = mount(<EditForm {...defaultProps} />);
    const dialog = wrapper.find(ScheduleDialog);
    act(() => {
      dialog.props().onDateSelect(null);
    });
    expect(defaultProps.onDateSelect).toBeCalled();
  });
});
