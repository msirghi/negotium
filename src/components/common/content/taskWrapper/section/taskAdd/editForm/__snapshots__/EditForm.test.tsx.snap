// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditForm should match the snapshot 1`] = `
<div
  className="makeStyles-root-1"
>
  <div
    className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
  >
    <div
      className="MuiInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-1pa58rh-MuiInputBase-root-MuiInput-root"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-nz481w-MuiInputBase-input-MuiInput-input"
        data-testid="tab-title-field"
        disabled={false}
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="New task"
        required={false}
        type="text"
        value="val"
      />
    </div>
  </div>
  <div>
    <div
      className="makeStyles-root-2"
      id="schedule-button"
      onClick={[Function]}
      style={
        Object {
          "border": "1px solid lightgrey",
          "color": "",
        }
      }
    >
      <div>
        Schedule
      </div>
    </div>
  </div>
</div>
`;
